### PR TITLE
online_config: allow return error when update config

### LINKED ISSUE #12909: replace Into with TryInto when convert ConfigValue

### PR DESCRIPTION
**Pull request by glorv**
Signed-off-by: glorv <glorvs@163.com>

### What is changed and how it works?

Issue Number: Close #12909

What's Changed:

```
Replace Into::into with TryInto::try_info to simplify the type conversion of ConfigValue
```

Currently, OnlineConfig depends on `From`/`Into` to convert between `ConfigValue` and the actual config type. For enum types that only accepts some specific string or integer values we have to define a enum variant like `ConfigValue::IOPriority(String)`, `ConfigValue::BlobRunMode(String)`. And the variants count increase linearly with new enum types. 

By changing the type conversion interface from `Into` to `TryInfo`, we can simplify this kind of type into `ConfigValue::String`. 

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

### Release note 

```release-note
None
```

### CODE REVIEW

**Comment by glorv** [comment]
@tabokie  @BusyJay  PTAL

File: components/tikv_util/src/config.rs [hunk 1]
```diff
@@ -1117,6 +1117,15 @@ impl<T> VersionTrack<T> {
         self.version.fetch_add(1, Ordering::Release);
     }
 
+    pub fn try_update<F>(&self, f: F) -> Result<(), Box<dyn std::error::Error>>
```
Comment (tabokie) [inline components/tikv_util/src/config.rs:1120]: Why not change the `update`?

File: components/resolved_ts/src/endpoint.rs [hunk 2]
```diff
@@ -579,7 +579,10 @@ where
     fn handle_change_config(&mut self, change: ConfigChange) {
         let prev = format!("{:?}", self.cfg);
         let prev_advance_ts_interval = self.cfg.advance_ts_interval;
-        self.cfg.update(change);
+        if let Err(e) = self.cfg.update(change).unwrap() {
```
Comment (tabokie) [inline components/resolved_ts/src/endpoint.rs:582]: What's with this `unwrap`?

**Review by tabokie — APPROVED** [review]

**Review by zhangjinpeng1987 — APPROVED** [review]
LGTM

**Comment by zhangjinpeng1987** [comment]
@glorv please check the ci.

File: components/engine_rocks/src/config.rs [hunk 3]
```diff
@@ -225,21 +225,22 @@ pub enum BlobRunMode {
...
-                m => panic!("expect: kNormal, kReadOnly or kFallback, got: {:?}", m),
-            }
+impl TryFrom<ConfigValue> for BlobRunMode {
+    type Error = String;
+    fn try_from(c: ConfigValue) -> Result<BlobRunMode, Self::Error> {
+        if let ConfigValue::String(s) = c {
+            Self::from_str(&s)
         } else {
-            panic!("expect: ConfigValue::BlobRunMode, got: {:?}", c);
+            panic!("expect: ConfigValue::String, got: {:?}", c);
```
Comment (BusyJay) [inline components/engine_rocks/src/config.rs:243]: Why not return error instead?

File: components/tikv_util/src/config.rs [hunk 4]
```diff
@@ -1130,6 +1133,24 @@ impl<T> VersionTrack<T> {
     }
 }
 
+/// CfgChanged is used to determine after call `VersionTrack::update`,
+/// whether the config is changed.
+pub trait CfgChanged: 'static {
+    fn changed(&self) -> bool;
+}
+
+impl CfgChanged for () {
```
Comment (BusyJay) [inline components/tikv_util/src/config.rs:1142]: This is very strange. Why not make the function always return `Result`?

File: components/engine_rocks/src/config.rs (hunk 3 above)
Comment (glorv) [reply components/engine_rocks/src/config.rs:243]: Normal logic can't enter this branch, so I think it's ok to just panic because this must be a bug then.

Comment (BusyJay) [reply components/engine_rocks/src/config.rs:243]: Is it possible that TiDB configures it with an int by SQL? Or can this be triggered by invalid client requests?

Comment (glorv) [reply components/engine_rocks/src/config.rs:243]: No. The input (either toml or json) will first be changed to a ConfigValue with the same type, if the type is not match, the first step will return an error. See: https://github.com/tikv/tikv/blob/a9c3e56552c803642f640e9b3fa8725aa3072400/src/config.rs#L3573-L3590

**Comment by glorv** [comment]
@BusyJay PTAL, thanks

**Review by BusyJay — APPROVED** [review]

File: components/cdc/src/endpoint.rs [hunk 5]
```diff
@@ -459,7 +462,7 @@ impl<T: 'static + RaftStoreRouter<E>, E: KvEngine> Endpoint<T, E> {
             "change" => ?change
         );
         // Update the config here. The following adjustments will all use the new values.
-        self.config.update(change.clone());
+        self.config.update(change.clone()).unwrap();
```
Comment (overvenus) [inline components/cdc/src/endpoint.rs:465]: Seems like it may panic.

**Pull request merged**